src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
-        return null;
+        verifyIndexIsValid(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        verifyIndexIsValid(index);
+        Node<T> nodeToRemove = getNode(index);
+        unlink(nodeToRemove);
+        size--;
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        while (currentNode != null) {
+            if (elementsAreEqual(currentNode.value, object)) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        verifyIndexIsValid(index);
+        return index <= size / 2
+                ? getNodeByIteratingFromBeginning(index) : getNodeByIteratingFromEnd(index);
+    }
+
+    private Node<T> getNodeByIteratingFromBeginning(int index) {
+        int counter = 0;
+        Node<T> currentNode = first;
+        while (counter != index) {
+            currentNode = currentNode.next;
+            counter++;
+        }
+        return currentNode;
+    }
+
+    private Node<T> getNodeByIteratingFromEnd(int index) {
+        int counter = size - 1;
+        Node<T> currentNode = last;
+        while (counter != index) {
+            currentNode = currentNode.prev;
+            counter--;
+        }
+        return currentNode;
+    }
+
+    private void unlink(Node<T> node) {
+        if (first == node && last == node) {
+            first = null;
+            last = null;
+            return;
+        }
+        if (node.prev == null) {
+            first = node.next;
+        } else if (node.next == null) {
+            last = node.prev;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+    }
+
+    private boolean verifyIndexIsValid(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(
+                    String.format("Invalid index passed: %d for size: %d", index, size));
+        }
+        return true;
+    }
+
+    private boolean elementsAreEqual(T first, T last) {
+        return (first == last) || ((first != null) && (first.equals(last)));
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.next = next;
+            this.prev = prev;
+        }
     }
 }
```

Comment:
file doesn't end with new line

Reply:
probably github automatically trims the last empty line, because it is visible in intellij idea, and if it was not there, the tests would not pass
![image](https://user-images.githubusercontent.com/86558116/236898895-f3288aeb-1058-4b95-8dd7-38de7f051f38.png)